salt/transport/zeromq.py

```diff
@@ -0,0 +1,654 @@
+            )
+
+        recon_delay = self.opts['recon_default']
+
+        if self.opts['recon_randomize']:
+            recon_delay = randint(self.opts['recon_default'],
+                                  self.opts['recon_default'] + self.opts['recon_max']
+                          )
+
+            log.debug("Generated random reconnect delay between '{0}ms' and '{1}ms' ({2})".format(
+                self.opts['recon_default'],
+                self.opts['recon_default'] + self.opts['recon_max'],
+                recon_delay)
+            )
+
+        log.debug("Setting zmq_reconnect_ivl to '{0}ms'".format(recon_delay))
+        self._socket.setsockopt(zmq.RECONNECT_IVL, recon_delay)
+
+        if hasattr(zmq, 'RECONNECT_IVL_MAX'):
+            log.debug("Setting zmq_reconnect_ivl_max to '{0}ms'".format(
+                self.opts['recon_default'] + self.opts['recon_max'])
+            )
+
+            self._socket.setsockopt(
+                zmq.RECONNECT_IVL_MAX, self.opts['recon_max']
+            )
+
+        if self.opts['ipv6'] is True and hasattr(zmq, 'IPV4ONLY'):
+            # IPv6 sockets work for both IPv6 and IPv4 addresses
+            self._socket.setsockopt(zmq.IPV4ONLY, 0)
+
+    def destroy(self):
+        if hasattr(self, '_stream'):
+            # TODO: Optionally call stream.close() on newer pyzmq? Its broken on some
+            self._stream.io_loop.remove_handler(self._stream.socket)
+            self._stream.socket.close(0)
+        self.context.term()
+
+    def __del__(self):
+        self.destroy()
+
+    # TODO: this is the time to see if we are connected, maybe use the req channel to guess?
+    @tornado.gen.coroutine
+    def connect(self):
+        if not self.auth.authenticated:
+            yield self.auth.authenticate()
+        self.publish_port = self.auth.creds['publish_port']
+        self._socket.connect(self.master_pub)
+
+    @property
+    def master_pub(self):
+        '''
+        Return the master publish port
+        '''
+        return 'tcp://{ip}:{port}'.format(ip=self.opts['master_ip'],
+                                          port=self.publish_port)
+
+    @tornado.gen.coroutine
+    def _decode_messages(self, messages):
+        '''
+        Take the zmq messages, decrypt/decode them into a payload
+        '''
+        messages_len = len(messages)
+        # if it was one message, then its old style
+        if messages_len == 1:
+            payload = self.serial.loads(messages[0])
+        # 2 includes a header which says who should do it
+        elif messages_len == 2:
+            payload = self.serial.loads(messages[1])
+        else:
+            raise Exception(('Invalid number of messages ({0}) in zeromq pub'
+                             'message from master').format(len(messages_len)))
+        ret = yield self._decode_payload(payload)
+        raise tornado.gen.Return(ret)
+
+    @property
+    def stream(self):
```

Comment:
My problem is very particular and seems unrelated to the regular workflow, so, lets leave it alone, at least while an issue regarding regular salt workflow comes up. I can just issue a test.ping(which I actually do need to) for what I'm currently working on...